### PR TITLE
ARROW-3199: [Plasma] File descriptor send and receive retries

### PR DESCRIPTION
**Pull request by pcmoritz**
An additional piece of eyes would be appreciated for this. It seems to solve our issue reported in the JIRA, but I'm not sure what the semantics of partial reads/writes is here (in particular, how are partial read/writes handled for ancillary data like file descriptors?).

found by cc @stephanie-wang

### CODE REVIEW

**Comment by wesm** [comment]
@pcmoritz is this the cause of

```
pyarrow/tests/test_plasma.py::test_plasma_list FAILED                                                                                                                                   [ 96%]
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> captured stderr >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
Allowing the Plasma store to use up to 0.1GB of memory.
Starting object store with directory /dev/shm and huge page support disabled
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> traceback >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>

    @pytest.mark.plasma
    def test_plasma_list():
        import pyarrow.plasma as plasma
    
        with plasma.start_plasma_store(
                plasma_store_memory=DEFAULT_PLASMA_STORE_MEMORY) \
                as (plasma_store_name, p):
            plasma_client = plasma.connect(plasma_store_name, "", 0)
    
            # Test sizes
            u, _, _ = create_object(plasma_client, 11, metadata_size=7, seal=False)
            l1 = plasma_client.list()
            assert l1[u]["data_size"] == 11
            assert l1[u]["metadata_size"] == 7
    
            # Test ref_count
            v = plasma_client.put(np.zeros(3))
            l2 = plasma_client.list()
            # Ref count has already been released
            assert l2[v]["ref_count"] == 0
            a = plasma_client.get(v)
            l3 = plasma_client.list()
>           assert l3[v]["ref_count"] == 1
E           assert 0 == 1

pyarrow/tests/test_plasma.py:820: AssertionError
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> entering PDB >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
> /home/wesm/code/arrow/python/pyarrow/tests/test_plasma.py(820)test_plasma_list()
-> assert l3[v]["ref_count"] == 1
```

Just seeing this locally

**Comment by pcmoritz** [comment]
Hmm, that sounds unlikely to me. What's the configuration you are running the test on, do you think I could reproduce it?

**Comment by wesm** [comment]
Running on Ubuntu 14.04, clang-6.0

**Comment by pcmoritz** [comment]
And which python?

File: cpp/src/plasma/fling.cc [hunk 1]
```diff
@@ -46,11 +48,31 @@ int send_fd(int conn, int fd) {
...
+        // errors, but sometimes I see a "Message too long" error number.
+        // Second, looping like this allows a client to potentially block the
+        // plasma store event loop which should never happen.
+        continue;
+      } else {
+        ARROW_LOG(INFO) << "Error in send_fd (errno = " << errno << ")";
+        return static_cast<int>(r);
+      }
+    } else if (r == 0) {
+      ARROW_LOG(INFO) << "Encountered unexpected EOF";
```
Comment (robertnishihara) [inline cpp/src/plasma/fling.cc]: In this case, we may need to return right? Will we infinite loop here?

In the past it looks like we were returning `0` here and indicating success.

Comment (pcmoritz) [reply cpp/src/plasma/fling.cc]: True, I fixed it!

Comment (robertnishihara) [reply cpp/src/plasma/fling.cc]: I'm still a little unclear on this case. Does `r == 0` mean the recipient hung up? In that case we return `0` and it's treated as success?

Comment (pcmoritz) [reply cpp/src/plasma/fling.cc]: Yeah, that seems to be the right behavior. Not there will also be a "Disconnected client" message in the main loop of the store to detect this.

**Pull request closed without merging**